exchanges/gateio/gateio_test.go

```diff
@@ -3301,10 +3301,45 @@ func getPairs(tb testing.TB, a asset.Item) currency.Pairs {
 func BenchmarkTimeInForceFromString(b *testing.B) {
 	for b.Loop() {
 		for _, tifString := range []string{gtcTIF, iocTIF, pocTIF, fokTIF} {
-			tif := timeInForceFromString(tifString)
-			if tif == order.UnknownTIF {
+			_, err := timeInForceFromString(tifString)
+			if err != nil {
 				b.Fatal(tifString)
 			}
 		}
 	}
 }
+
+func TestTimeInForceFromString(t *testing.T) {
+	t.Parallel()
+	_, err := timeInForceFromString("abcdef")
+	assert.ErrorIs(t, err, order.ErrUnsupportedTimeInForce)
+
+	for k, v := range map[string]order.TimeInForce{gtcTIF: order.GoodTillCancel, iocTIF: order.ImmediateOrCancel, pocTIF: order.PostOnly, fokTIF: order.FillOrKill} {
+		tif, err := timeInForceFromString(k)
+		assert.NoError(t, err)
+		assert.Equal(t, v, tif)
+	}
+}
+
+func TestGetTypeFromTimeInForce(t *testing.T) {
+	t.Parallel()
+	typeResp := getTypeFromTimeInForce("gtc", 0)
+	assert.Equal(t, order.Limit, typeResp)
+
+	typeResp = getTypeFromTimeInForce("ioc", 0)
+	assert.Equal(t, order.Market, typeResp, "should be market order")
+
+	typeResp = getTypeFromTimeInForce("poc", 123)
+	assert.Equal(t, order.Limit, typeResp, "should be limit order")
+
+	typeResp = getTypeFromTimeInForce("fok", 0)
+	assert.Equal(t, order.Market, typeResp, "should be market order")
+}
+
+func TestTimeInForceString(t *testing.T) {
+	t.Parallel()
+	assert.Empty(t, timeInForceString(order.UnknownTIF))
+	for _, valid := range validTimesInForces {
+		assert.Equal(t, valid.String, timeInForceString(valid.TimeInForce))
+	}
```

Comment:
🐒 There are not many values, and this approach to testing risks just carrying a bug over straight from the map into the test.
Can we hardcode the expected values here?
It's verbose but sometimes that's the point of testing.

Reply:
It's just to minimize hard coding ... this implementation is much efficient so I prefer to leave it like this

Reply:
Okay. I still don't agree don't agree on either of your reasons.
The point of tests is not to be efficient.
The test *should* be divergent from the source data, and duplicate the hardcoding, because otherwise if someone accidentally deletes a line from the source, the test won't catch it.

But this was a nitpick, so it was optional. Resolving.